include/seqan3/search/algorithm/search_result_range.hpp

```diff
@@ -0,0 +1,320 @@
+    bool first_call_of_begin{true};
+};
+
+/*!\brief The iterator of seqan3::detail::search_result_range.
+ * \implements std::input_iterator
+ */
+template <typename search_algorithm_t, std::ranges::view query_range_t>
+//!\cond
+    requires std::ranges::forward_range<query_range_t> &&
+             std::invocable<search_algorithm_t, std::ranges::range_reference_t<query_range_t>>
+//!\endcond
+class search_result_range<search_algorithm_t, query_range_t>::search_result_range_iterator
+{
+private:
+    //!\brief The internal result buffer iterator.
+    using result_buffer_iterator = std::ranges::iterator_t<result_buffer_t>;
+public:
+    /*!\name Associated types
+     * \{
+     */
+    //!\brief Type for distances between iterators.
+    using difference_type = std::ptrdiff_t;
+    //!\brief Value type of container elements.
+    using value_type = result_buffer_value_t;
+    //!\brief Use reference type defined by container.
+    using reference = result_buffer_value_t &;
+    //!\brief Pointer type is pointer of container element type.
+    using pointer = std::add_pointer_t<value_type>;
+    //!\brief Sets iterator category as input iterator.
+    using iterator_category = std::input_iterator_tag;
+    //!\}
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr search_result_range_iterator() noexcept = default; //!< Defaulted.
+    constexpr search_result_range_iterator(search_result_range_iterator const &) noexcept = default; //!< Defaulted.
+    constexpr search_result_range_iterator(search_result_range_iterator &&) noexcept = default; //!< Defaulted.
+    constexpr search_result_range_iterator & operator=(search_result_range_iterator const &) noexcept = default; //!< Defaulted.
+    constexpr search_result_range_iterator & operator=(search_result_range_iterator &&) noexcept = default; //!< Defaulted.
+    ~search_result_range_iterator() = default; //!< Defaulted.
+
+    /*!\brief Construct from the associated search result range.
+     * \param[in] range The associated search result range.
+     * \param[in] first_call_of_begin A flag indicating whether this is the first call of begin.
+     *
+     * \details
+     *
+     * Initialises the iterator with the associated search result range. If `first_call_of_begin` is `true`,
+     * the search algorithm is invoked on the first query.
+     */
+    constexpr search_result_range_iterator(search_result_range & range, bool const first_call_of_begin) :
+        range_ptr{& range},
+        at_end(range.at_end())
+    {
+        if (first_call_of_begin)
+            fetch_next_query_results();
+    }
+    //!\}
+
+    /*!\name Access operators
+     * \{
+     */
+
+    /*!\brief Access the pointed-to element.
+     * \returns A reference to the current element.
+     */
+    //!\brief Returns the current search result.
+    reference operator*() const noexcept
+    {
+        return *current_buffer_it;
```

Comment:
The iterator is an input iterator and does need to be copyable since C++20 anymore as long as it does not model CPP17Iterator, which is not required in this case. In fact, it is discouraged to copy input iterators in general because of the unclear behaviour. The stream iterators indeed store a copy of the element inside of the iterator. But that was not ideal before, when we had to think about possibly copying this iterator. I am not sure about the range library and if they updated the concepts already. In this case it would be fine to mark this iterator move-only and the iterator could store all information making the range a borrowed range. But that wasn't like so when I wrote the version and thus it was better to keep the data in the range and iterator just pointing to it in case someone was copying it for what ever reason. I believe a lot of stuff breaks once we enforce this change. And I think there is no general benefit of changing now everything. So, once we know that current input iterator concepts are not required to be copyable anymore we can make all input iterators move-only and the code above should be fine.

Reply:
Ok, thanks for the explanation. I will make a card for this so we don't forget to check the concepts.